### PR TITLE
ALP-3164 Eth entry connection changes

### PR DESCRIPTION
**Pull request by pchy10**

### CODE REVIEW

**Comment by larryob** [comment]
Why are we passing fake rpc urls here: https://github.com/AffineLabs/defi-sdk/blob/b3164d60d3d8c1b9fb27034efa20021a892e04d8/src/frontend/Account.ts#L253-L265

If the rpc urls aren't being used, could we just get rid of them?

**Comment by pchy10** [comment]
> 

@larryob This has been resolved with other required changes. I pushed the final commit a while ago. Please take a look again.

File: src/core/constants.ts [hunk 1]
```diff
@@ -1,10 +1,16 @@
...
 
 export const DEFAULT_WALLET = "magic";
 
 // The chain id is a hexadecimal string preceeded by "0x"
-const rawId = process.env.CHAIN_ID || "80001";
-export const CHAIN_ID = `0x${Number(rawId).toString(16)}`;
+export const DEFAULT_RAW_CHAIN_ID: AllowedChainId = process.env.CHAIN_ID
+  ? (parseInt(process.env.CHAIN_ID) as AllowedChainId)
+  : 80001;
+// export const DEFAULT_CHAIN_ID = `0x${Number(rawId).toString(16)}`;
```
Comment (larryob) [inline src/core/constants.ts:10]: You can delete this.

File: src/frontend/Account.ts [hunk 2]
```diff
@@ -42,79 +43,61 @@ class Account {
...
-      // walletConnect doesn't require window.ethereum to be present always
-      const provider = (await getExternalProvider(walletType)) as WalletConnectProvider;
-
-      if (provider) this.walletConnectProvider = provider;
-
-      if (provider) {
-        walletProvider = new ethers.providers.Web3Provider(provider as ethers.providers.ExternalProvider);
-      }
+    } else {
+      walletProvider = await getWeb3Provider(walletType, chainId);
```
Comment (larryob) [inline src/frontend/Account.ts:66]: Why do we need to pass the chainId in order to get the provider?

Don't all the wallets (besides Magic) have something like `window.ethereum` that will provide the chainId for you?

I see [here](https://github.com/AffineLabs/defi-sdk/blob/0b22a1ca5f5b9e50a1c54d360b543097ae516ee8/src/frontend/wallets.ts#L70-L91) that for coinbase and wallet connect you're taking our alchemy url and using that as a provider. Is that necessary?

File: src/frontend/Account.ts [hunk 3]
```diff
@@ -142,16 +125,19 @@ class Account {
    */
   async disconnect(walletType: AllowedWallet): Promise<void> {
     if (walletType === "magic" && this.magic?.user) await this.magic.user.logout();
-    if (this.walletConnectProvider?.disconnect) this.walletConnectProvider.disconnect();
+    if (walletType === "walletConnect" && this.walletProvider) {
+      const walletConnectProvider = this.walletProvider.provider as WalletConnectProvider;
+      await walletConnectProvider.disconnect();
+    }
```
Comment (larryob) [inline src/frontend/Account.ts:131]: Can you leave a comment explaining why this is necessary in the case of walletconnect?

File: src/frontend/Account.ts [hunk 4]
```diff
@@ -235,59 +220,31 @@ class Account {
...
-      },
-      rpcUrls: ["https://polygon-rpc.com"],
-      blockExplorerUrls: ["https://polygonscan.com"],
-    };
-
-    const params = CHAIN_ID === "0x89" ? mainnetParams : testnetParams;
-    return params;
+  async getChainId(): Promise<string | undefined> {
+    if (!this.walletProvider) return;
+    return await this.walletProvider.getNetwork().then(network => network.chainId.toString());
```
Comment (larryob) [inline src/frontend/Account.ts:225]: The `await` is not needed.
Also I think this fits the style of the codebase more: 

```typescript
const {chainId} = await this.walletProvider.getNetwork();
return chainId.toString();
```

File: src/frontend/Account.ts [hunk 5]
```diff
@@ -235,59 +220,31 @@ class Account {
...
 
-    const ethProvider = await getExternalProvider(wallet ?? this.walletType);
-    console.log("Eth provider on switchWalletToAllowedNetwork", wallet, ethProvider);
+    const _provider = await getWeb3Provider(walletType, chainId);
 
-    const provider = new ethers.providers.Web3Provider(ethProvider as ethers.providers.ExternalProvider);
+    if (!_provider) {
+      throw new Error("Provider is not available");
+    }
+    // We have to pass "any" if we want to change networks. See https://github.com/ethers-io/ethers.js/issues/1107
```
Comment (larryob) [inline src/frontend/Account.ts:245]: Move this comment into `getWeb3Provider`.

File: src/frontend/Account.ts [hunk 6]
```diff
@@ -297,24 +254,33 @@ class Account {
          * case - 4902 indicates that the chain has not been added to MetaMask.
          * @see https://docs.metamask.io/guide/rpc-api.html#usage-with-wallet-switchethereumchain
          */
-        await provider.send("wallet_addEthereumChain", [{ ...this._getNetworkParams() }]);
+        await _provider.send("wallet_addEthereumChain", [
+          { ...NETWORK_PARAMS[chainId], chainId: getChainIdFromRaw(chainId) },
```
Comment (larryob) [inline src/frontend/Account.ts:258]: This actually means that if the user adds this network to their wallet, then they will always use our rpc url (when connected to the given network), even when not using our app.

We should definitely change this in the future.

**Review by larryob — COMMENTED** [review]
Looking good, I left some comments.

File: src/frontend/Account.ts (hunk 2 above)
Comment (pchy10) [reply src/frontend/Account.ts:66]: No, both WalletConnect & Magic don't inject providers (like `window.ethereum`) into the browser. That's why we need to provide our RPC URLs to this kind of wallet.

File: src/frontend/Account.ts (hunk 6 above)
Comment (pchy10) [reply src/frontend/Account.ts:264]: I can't find any suitable public RPC URLs for `ethereum` and `goerli`. But I changed the RPC URLs anyway. Let me know if you have more suitable RPC URLs to add to the metamask.

Comment (larryob) [reply src/frontend/Account.ts:273]: The urls look good.

**Review by larryob — APPROVED** [review]
Looks good, thanks.

**Pull request merged**